src/main/scala/is/hail/annotations/UnsafeAnnotations.scala

```diff
@@ -0,0 +1,402 @@
+  }
+
+  private def putFloat(value: Float, offset: Int) {
+    assert((offset & 0x3) == 0, s"invalid float offset: $offset")
+    if (debug) println(s"offset $offset writing float $value")
+
+    Platform.putFloat(mem, absolute(offset), value)
+  }
+
+  private def putDouble(value: Double, offset: Int) {
+    assert((offset & 0x7) == 0, s"invalid double offset: $offset")
+    if (debug) println(s"offset $offset writing double $value")
+    Platform.putDouble(mem, absolute(offset), value)
+  }
+
+  private def putByte(value: Byte, offset: Int) {
+    if (debug) println(s"offset $offset writing byte $value")
+    Platform.putByte(mem, absolute(offset), value)
+  }
+
+  private def putBinary(value: Array[Byte], offset: Int) {
+    var cursor = appendPointer
+
+    // write the offset
+    putInt(appendPointer - offset, offset)
+
+    val totalSize = 4 + value.length
+    if (debug)
+      println(s"putting binary ${ value } at offset $offset with shift ${ appendPointer - offset }, totSize=$totalSize")
+
+    appendPointer = UnsafeAnnotations.roundUpAlignment(totalSize + appendPointer)
+    reallocate(appendPointer)
+
+    putInt(value.length, cursor)
+
+    cursor += 4
+
+    Platform.copyMemory(value, Platform.BYTE_ARRAY_OFFSET, mem, absolute(cursor), value.length)
+  }
+
+  private def putArray(value: Iterable[_], offset: Int, elementType: Type) {
+
+    val missingBytes = UnsafeAnnotations.missingBytes(value.size)
+    val eltSize = UnsafeAnnotations.arrayElementSize(elementType)
+
+    var cursor = appendPointer
+
+    // write the offset
+    putInt(appendPointer - offset, offset)
+
+    val totalSize = UnsafeAnnotations.roundUpAlignment(
+      UnsafeAnnotations.roundUpAlignment(4 + missingBytes) + value.size * eltSize)
+
+    if (debug)
+      println(s"putting array ${ value.toSeq } at offset $offset with shift ${ appendPointer - offset }, totSize=$totalSize")
+
+    appendPointer = totalSize + appendPointer
+    reallocate(appendPointer)
+    putInt(value.size, cursor)
+
+    cursor += 4
+    val missingBitStart = cursor
+
+    cursor += missingBytes
+
+    // align the element
+    cursor = UnsafeAnnotations.roundUpAlignment(cursor)
+
+    // TODO: specialize to numeric types. Non-nullable types make this easier, too
+    var i = 0
+    val iter = value.iterator
+    while (i < value.size) {
+      val elt = iter.next()
+      if (debug)
+        println(s"array index $i: writing value $elt (${ elementType.toPrettyString(compact = true) }) to offset $cursor")
+      if (elt == null) {
+        val intIndex = missingBitStart + ((i >> 5) << 2)
+        val bitShift = i & 0x1f
+        val oldBits = readInt(intIndex)
+        val newBits = oldBits | (0x1 << bitShift)
+        putInt(newBits, intIndex)
+      } else {
+        val shift = put(elt, cursor, elementType)
+      }
+      cursor += eltSize
+      i += 1
+    }
+  }
+
+  private def putStruct(value: Row, offset: Int, struct: TStruct) {
```

Comment:
I can't quite figure out how this code works.  I expected putStruct to allocate the space for the struct, and then put each field, with arrays and things being added after the struct.  You write each field, I just don't see the allocate.  Am I missing something?